presto-main/src/main/java/com/facebook/presto/util/Lazy.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.util;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public class Lazy<T>
+        implements Supplier<T>
+{
+    private T value;
+    private volatile Supplier<T> init;
+
+    public Lazy(Supplier<T> init)
+    {
+        reset(init);
+    }
+
+    public Lazy<T> reset(Supplier<T> init)
+    {
+        this.init = requireNonNull(init, "init is null");
+        return this;
+    }
+
+    public synchronized Lazy<T> reset(T value)
+    {
+        this.value = value;
+        this.init = null;
+        return this;
+    }
```

Comment:
Do we use these methods anywhere outside of the constructor?

Reply:
currently no, I will simplify the implementation